core/kernel/dt_driver.c

```diff
@@ -257,3 +408,293 @@ TEE_Result dt_driver_probe_device_by_node(const void *fdt, int nodeoffset,
 
 	return TEE_ERROR_ITEM_NOT_FOUND;
 }
+
+static TEE_Result process_probe_list(const void *fdt)
+{
+	struct dt_driver_probe *elt = NULL;
+	struct dt_driver_probe *prev = NULL;
+	unsigned int __maybe_unused loop_count = 0;
+	unsigned int __maybe_unused deferral_loop_count = 0;
+	bool __maybe_unused one_deferred = false;
+	bool all_deferred = false;
+
+	do {
+		loop_count++;
+		FMSG("Probe loop %u after %u for deferral(s)", loop_count,
+		     deferral_loop_count);
+
+		/* Hack here for TRACE_DEBUG messages on probe list elements */
+		if (TRACE_LEVEL >= TRACE_FLOW)
+			print_probe_list(fdt);
+
+		if (TAILQ_EMPTY(&dt_driver_probe_list))
+			return TEE_SUCCESS;
+
+		/*
+		 * Probe from current end to top. Deferred probed node are
+		 * pushed back after current tail for the next probe round.
+		 * Reset probe result flags and see status after probe round.
+		 */
+		one_deferred = false;
+		all_deferred = true;
+		added_node = false;
+
+		TAILQ_FOREACH_REVERSE_SAFE(elt, &dt_driver_probe_list,
+					   dt_driver_probe_head, link, prev) {
+			TAILQ_REMOVE(&dt_driver_probe_list, elt, link);
+
+			switch (probe_driver_node(fdt, elt)) {
+			case TEE_SUCCESS:
+				all_deferred = false;
+				break;
+			case TEE_ERROR_DEFER_DRIVER_INIT:
+				one_deferred = true;
+				break;
+			default:
+				/* We don't expect error return codes */
+				assert(0);
+			}
+		}
+
+		if (one_deferred)
+			deferral_loop_count++;
+
+	} while (added_node || !all_deferred);
+
+	EMSG("Panic on unresolved dependencies after %u rounds, %u deferred:",
+	     loop_count, deferral_loop_count);
+
+	TAILQ_FOREACH(elt, &dt_driver_probe_list, link)
+		EMSG("- %s on node %s", elt->dt_drv->name,
+		     fdt_get_name(fdt, elt->nodeoffset, NULL));
+
+	panic();
+}
+
+static bool candidate_is_list_elt(struct dt_driver_probe *candidate,
```

Comment:
I can't make sense of this function name. How about `driver_probe_compare()` instead?